morph_tool/morphdb.py

```diff
@@ -0,0 +1,346 @@
+            def get_repair_attr(morph, attr):
+                if attr == 'axon_inputs':
+                    return 'axon_sources', {'axoninput': getattr(morph, attr)}
+                else:
+                    return attr, getattr(morph, attr)
+            data = {'neurondb':
+                    {'listing': {
+                        'morphology': [{
+                            'name': morph.name,
+                            'mtype': morph.mtype_no_subtype,
+                            'msubtype': morph.msubtype,
+                            'layer': morph.layer,
+                            'repair': dict(
+                                get_repair_attr(morph, attr)
+                                for attr in BOOLEAN_REPAIR_ATTRS + ['axon_inputs']
+                            )
+                        } for morph in self.df.itertuples()]}
+                     }}
+            with output_path.open('w') as fd:
+                fd.write(xmltodict.unparse(data, pretty=True))
+        else:
+            raise ValueError(f'Unsupported neurondb extensions ({ext}).'
+                             ' Should be one of: (xml,csv,dat)')
+        return output_path
+
+    def features(self, config: Dict, n_workers=1):
+        '''Returns a dataframe containing morphometrics and neurondb information
+
+        Args:
+            config: a NeuroM morph_stas config.
+                See https://neurom.readthedocs.io/en/latest/morph_stats.html for more information
+
+        Returns:
+            A jointure dataframe between `neurom.stats.extract_dataframe` and `self.df`
+
+        Raises:
+            ValueError: if `self.df` has undefined (ie. None) paths
+        '''
+        missing_morphs = self.df[self.df.path.isnull()].name.values
+        if missing_morphs:
+            raise ValueError(
+                f'DataFrame has morphologies with undefined filepaths: {missing_morphs}')
+
+        df = self.df.copy().reset_index()
+        df.columns = pd.MultiIndex.from_product((["neuron"], df.columns.values))
+        paths = df['neuron', 'path']
+
+        stats = extract_dataframe(paths, config, n_workers).drop(columns='name', level=1)
+        return df.join(stats, how='inner').drop(columns=('neuron', 'index'))
+
+    def check_files_exist(self):
+        '''Raises if `self.df.path` has None values or non existing paths'''
+        missing_morphs = self.df[self.df.path.isnull()].name.values
+        if missing_morphs:
+            raise ValueError(
+                f'DataFrame has morphologies with undefined filepaths: {missing_morphs}')
+
+        for path in self.df.path:
+            if not path.exists():
+                raise ValueError(f'Non existing path: {path}')
+
+    def __add__(self, other):
+        obj = MorphDB()
+        if isinstance(other, MorphDB):
+            obj.df = pd.concat([self.df, other.df])
+        elif isiterable(other):
+            seq = list(other)
+            if seq and isinstance(seq[0], MorphInfo):
```

Comment:
I'm not fully convinced that allowing to add together `MorphDB` and a list of `MorphInfo` is intuitive (I think that the two operands should usually belong to the same type, or to compatible types).
However I can change my mind. Maybe do you know other similar examples in this or other libraries?

Reply:
I agree. I'll make the `MorphDB` ctor accept a sequence of `MorphInfo` object.
This way adding a list of MorphInfo and a MorphDB object will be:

```python
db2 = db1 + MorphDB(morph_info_seq)
```

Reply:
If useful, you could add also an `update` method that can accept an iterable (it's not unusual to update a collection with an iterable).

Reply:
I think for the update `db += MorphDB(iterable)` is good enough, no ?